sound/soc/sof/debug-dsp-ops.c

```
@@ -8,6 +8,33 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+static int sof_dsp_ops_boot_firmware(struct snd_sof_dev *sdev)
+{
+	const char *fw_filename;
+	int ret;
+
+	sdev->first_boot = true;
+
+	if (!sdev->test_profile.fw_path || !sdev->test_profile.fw_name) {
+		dev_dbg(sdev->dev, "Invalid firmware path or filename\n");
+		return -EINVAL;
+	}
+
+	fw_filename = kasprintf(GFP_KERNEL, "%s/%s", sdev->test_profile.fw_path,
+				sdev->test_profile.fw_name);
+
+	/* load firmware */
+	ret = snd_sof_load_firmware(sdev, fw_filename);
```

How this will pass our test in `snd_sof_load_firmware_raw()` ?
```c
	/* Don't request firmware again if firmware is already requested */
	if (sdev->basefw.fw)
		return 0;
```
You likely need to release the firmware, wipe all information about modules, possibly drop loaded external libraries as well.
Likely the topology needs to be re-verified to see if the firmware have all the modules that is needed, so destroy the card?

---

see my comment below. With the debug option selected nothing is loaded at boot

---

right, but you can load it once, is this by design?
And on consecutive tries no one would know if the new firmware was ignored.

---

The problem is if someone wants to load a new firmware, they need to make sure the DSP is powered off first and the previous FW is unloaded before attempting to boot again. Im not sure we should handle these in the kernel.

---

Should we tell the user that what is tried is going to be a NOP and the previously loaded kernel is going to remain?

---

We need to be clear on what we mean by 'new firmware'. if someone overrides an existing firmware in /lib/firmware, they will face all kinds of caching issues. We cannot really add a new firmware with SSH and hope that it's used immediately, that's really asking for trouble.

The assumption should be that you test the firmware that was available at boot.

---

> The assumption should be that you test the firmware that was available at boot.

"Only at boot" and "all kinds of caching issues" sound a bit too extreme to me. I mean the requirement to reboot to test a different firmware version sounds too extreme? I vaguely remember some `request_firmware()` mis-use and caching bug a couple years ago but besides that one, https://github.com/thesofproject/sof-test/blob/main/tools/kmod/sof_remove.sh has been working just fine to test new firmware. Granted this was before libraries but  why couldn't this achieve the same? At least when libraries are not used.

I this requires some mass `rmmod` like `sof_remove.sh` does and like @ujfalusi seems to request then it looks good to me.

BTW do we care about libraries here? I thought this was meant to debug early DSP boot issues. Once those are fixed, just debug the firmware as usual/before?

cc: @andyross

---

> Should we tell the user that what is tried is going to be a NOP and the previously loaded kernel is going to remain?

NEVER, ever silently discard user input. That's sending people who already losing hair debugging problems straight to the asylum. If you must discard user input for some reason, then it's fine but return every major error code you can (for scripts), and log at the highest level every error message you can - in bold face and capitals (for both scripts and humans)

---

OK I've fixed this to boot default FW is the user chooses not to pass the test firmware file